core/src/main/scala/kafka/log/Log.scala

```diff
@@ -2124,13 +2132,15 @@ class Log(@volatile private var _dir: File,
           time = time,
           initFileSize = initFileSize,
           preallocate = config.preallocate))
-        updateLogEndOffset(newOffset)
         leaderEpochCache.foreach(_.clearAndFlush())
 
-        producerStateManager.truncate()
-        producerStateManager.updateMapEndOffset(newOffset)
+        // Clear producer state prior to adjusting log start and end offsets
+        // in case the first unstable offset is invalidated by the truncation.
+        producerStateManager.truncateFullyAndStartAt(newOffset)
         maybeIncrementFirstUnstableOffset()
+
         updateLogStartOffset(newOffset)
+        updateLogEndOffset(newOffset)
```

Comment:
This is an existing issue. updateLogEndOffset() makes the assumption that newOffset is nextOffset since it uses nextOffsetMetadata directly. This can be error prone in the future. While you are at this, perhaps we could remove the input param from updateLogEndOffset()?

Reply:
Actually, this is not an issue on second look.